src/ConstructionBase.jl

```diff
@@ -43,13 +43,24 @@ end
 
 getproperties(o::NamedTuple) = o
 getproperties(o::Tuple) = o
-@generated function getproperties(obj)
-    fnames = fieldnames(obj)
-    fvals = map(fnames) do fname
-        Expr(:call, :getproperty, :obj, QuoteNode(fname))
+if VERSION >= v"1.7"
+    function getproperties(obj)
+        fnames = propertynames(obj)
+        NamedTuple{fnames}(getproperty.(Ref(obj), fnames))
+    end
+else
+    @generated function getproperties(obj::T) where {T}
+        if which(propertynames, Tuple{T}).sig != Tuple{typeof(propertynames), Any}
```

Comment:
Interesting, technically only pure functions are allowed in `@generated`. But this might be ok in practice?